File: backend/models/survey.py
from typing import Dict, List, Optional
from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field

    
class PreTaskSurveyRequest(BaseModel):
    participant_id: str
    field_of_study: str
    confidence_level: int = Field(ge=1, le=5)
    use_of_AI: bool
    which_AI: Optional[str] = None


class SurveySession(BaseModel):
    participant_id: str
    started_at: datetime
    completed_at: Optional[datetime] = None
    survey_id: str
    current_task_index: int = 1

    model_config = ConfigDict(
        arbitrary_types_allowed=True,
    )

class TaskResultRequest(BaseModel):
    participant_id: str
    task_index: int = Field(ge=1, le=4)
    time_taken: int = Field(ge=1, le=7*60) # in seconds
    finished_within_time: bool
    code_files: List[Dict[str, str]]
    run_output: Optional[str] = None

class PostTaskSurveyRequest(BaseModel):
    participant_id: str
    task_index: int = Field(ge=1, le=4)
    finished_within_time: bool
    difficult_to_fix: int = Field(ge=1, le=5)
    helpful_understand: int = Field(ge=1, le=5)
    helpful_fix: int = Field(ge=1, le=5)
    thought_process: str
    feedback: str

class SuSItem(BaseModel):
    q7: int = Field(ge=1, le=5)
    q8: int = Field(ge=1, le=5)
    q9: int = Field(ge=1, le=5)
    q10: int = Field(ge=1, le=5)
    q11: int = Field(ge=1, le=5)
    q12: int = Field(ge=1, le=5)
    q13: int = Field(ge=1, le=5)
    q14: int = Field(ge=1, le=5)
    q15: int = Field(ge=1, le=5)
    q16: int = Field(ge=1, le=5)

class OverallSurveyRequest(BaseModel):
    participant_id: str
    sus : SuSItem
    future_use_likelihood: int = Field(ge=1, le=5)
    preferred_feedback_style: str
    preferred_feedback_reason: str
    other_comments: Optional[str] = None

File: my-coding-coach/src/lib/api/survey.ts
import { api } from "../http";
import { SurveyType } from "../survey/types";

// Survey types
export interface PreSurveyData {
    participant_id: string;
    field_of_study: string;
    confidence_level: number;
    use_of_AI: boolean;
    which_AI?: string;
}

export interface TaskResultData {
    participant_id: string;
    survey_type: string;
    task_index: number;
    time_taken: number;
    finished_within_time: boolean;
    code_files: Array<{ name: string; content: string; language: string }>;
    run_output?: string;
}

export interface PostTaskSurveyData {
    participant_id: string;
    survey_type: string;
    task_index: number;
    finished_within_time: boolean;
    difficult_to_fix: number;
    helpful_understand: number;
    helpful_fix: number;
    thought_process: string;
    feedback: string;
}

export interface OverallSurveyData {
    participant_id: string;
    survey_type: string;
    sus: {
        q7: number; q8: number; q9: number; q10: number; q11: number;
        q12: number; q13: number; q14: number; q15: number; q16: number;
    };
    future_use_likelihood: number;
    preferred_feedback_style: string;
    preferred_feedback_reason: string;
    other_comments?: string;
}

// Survey API functions

export const startSurvey = async (surveyType: SurveyType) => {
    console.log('surveyType', surveyType);
    return api.post('/api/survey/start', { survey_type: surveyType });
};

export const submitPreSurvey = async (data: PreSurveyData) => {
    return api.post('/api/survey/pre-task', data);
};

export const saveTaskResult = async (data: TaskResultData) => {
    return api.post('/api/survey/task-result', data);
};

export const submitPostTaskSurvey = async (data: PostTaskSurveyData) => {
    return api.post('/api/survey/post-task', data);
};

export const submitOverallSurvey = async (data: OverallSurveyData) => {
    return api.post('/api/survey/overall', data);
};